### PR TITLE
U30 xbutil2 enhancement

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
CR : https://jira.xilinx.com/browse/CR-1087398

XRT needs to monitor more metrics for the edge stack running on U30 PS. We should at the minimum enhance xbutil query to
report the following:

Xbutil2 Enhancement

1. Memory Report
      Show all buffers information in memory report
               -  xclGetHostBO
               -  xclMapBO
               -  xclUnmapBO
               -  xclFreeBO
               
  ```
Soft Kernel Memory Status
    HostBO Count      MapBO Count     UnmapBO Count   FreeBO Count    
    0                 0               0               0
```     

2. SKD report
       Show all the skd processes running
              - Show how many times a SKD triggered
              - Metrics on how many times SKDs have crashed or have returned nonzero error code
                    - Success Count
                    - Error Count 
                    - Crash Count 
```
 Success_Count   Error_Count     Crash_Count
 608             0               0          
 0               0               0          
 0               0               0          
 0               0               0          
 0               0               0          
 0               0               0          
 0               0               0   
```       

Sample xbutil2 output : 
**xbutil  examine -r memory -d 5f:00.1**
```
Memory Information

  Memory Topology
         Tag         Type        Temp(C)  Size      Base Address    
    [ 0] DDR[0]      MEM_DRAM    N/A      1023 MB   0x800000000     

  Memory Status
         Tag         Type        Size    Mem Usage       BO count
    [ 0] DDR[0]      MEM_DRAM    1 GB    0 Byte          0       

  Soft Kernel Memory Status
    HostBO Count      MapBO Count     UnmapBO Count   FreeBO Count    
    0                 0               0               0              

```
JSON Output for Memory Status
```
"memories": [
    {
        "type": "MEM_DRAM",
        "tag": "DDR[0]",
        "enabled": "true",
        "base_address": "0x800000000",
        "range_bytes": "0x3fffc000",
        "extended_info": {
            "usage": {
                "allocated_bytes": "0",
                "buffer_objects_count": "0"
            }
        }
    }
],
"sk_memories": {
    "sk_hbo": "42",
    "sk_mapbo": "42",
    "sk_unmapbo": "29",
    "sk_freebo": "27"
},
```

**xbutil  examine -r dynamic-regions -d 5f:00.1**
```
Compute Units
  PL Compute Units
    Index   Name                          Base_Address    Usage   Status  
    0       scaler:scaler_1               0x1400000       0       (IDLE)  
    1       lookahead:lookahead_1         0x1440000       0       (IDLE)  
    2       lookahead:lookahead_2         0x1450000       0       (IDLE)  
    3       decoder:decoder_1             0x1451000       0       (IDLE)  
    4       encoder:encoder_1             0x1452000       0       (IDLE)  
    5       VCU:VCU_1                     0x1453000       0       (IDLE)  

  PS Compute Units
    Index   Name                          Base_Address    Usage   Status      Success_Count   Error_Count     Crash_Count     
    0       kernel_vcu_decoder_0          0x0             608     (IDLE)      608             0               0               
    1       kernel_vcu_decoder_1          0x0             0       (IDLE)      0               0               0               
    2       kernel_vcu_decoder_2          0x0             0       (IDLE)      0               0               0               
    3       kernel_vcu_decoder_3          0x0             0       (IDLE)      0               0               0               
    4       kernel_vcu_decoder_4          0x0             0       (IDLE)      0               0               0               
    5       kernel_vcu_decoder_5          0x0             0       (IDLE)      0               0               0               
    6       kernel_vcu_decoder_6          0x0             0       (IDLE)      0               0               0               
    7       kernel_vcu_decoder_7          0x0             0       (IDLE)      0               0               0               
            
```
JSON output for a single PS compute Unit:
```
                        {
                            "name": "kernel_vcu_decoder_1",
                            "base_address": "0x0",
                            "usage": "0",
                            "type": "PS",
                            "succ_return": "0",
                            "err_return": "0",
                            "crsh_return": "0",
                            "status": {
                                "bit_mask": "0x4",
                                "bits_set": [
                                    "IDLE"
                                ]
                            }
                        },

```

### CODE REVIEW

**Comment by saifuddin-xilinx** [comment]
The previous PR #5470  has been created on origin/u30 branch. But currently U30 branch is upgraded to 2021.1 branch.
Hence created a new PR targeted to 2021.1 and closing this PR.

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by AShivangi — COMMENTED** [review]
```
  Memory Topology
         Tag         Type        Temp(C)  Size      Base Address    
    [ 0] DDR[0]      MEM_DRAM    N/A      1023 MB   0x800000000 
```
IMO should be:
```
  Memory Topology
   Index   Tag         Type        Temp(C)  Size      Base Address    
     0    DDR[0]      MEM_DRAM    N/A      1023 MB   0x800000000 
```

Can you also add the json output for @rozumx and I to review the nomenclature?

Thank you

**Comment by saifuddin-xilinx** [comment]
@AShivangi I have updated the JSON output in the PR description. Please check.

File: src/runtime_src/core/common/drv/include/kds_core.h [hunk 1]
```diff
@@ -57,12 +57,27 @@ struct kds_ctx_info {
...
+        u32                       freebo_cnt;
+};
+
 struct kds_scu_mgmt {
 	struct mutex		  lock;
 	int			  num_cus;
 	u32			  status[MAX_CUS];
 	u32			  usage[MAX_CUS];
 	char			  name[MAX_CUS][32];
+	struct scu_usages	  usages_stats[MAX_CUS];
```
Comment (houlz0507) [inline src/runtime_src/core/common/drv/include/kds_core.h:79]: Could you Merge "usage" to scu_usages structure?

Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_core.h:79]: Yes. We can merge.

Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/include/kds_core.h:79]: Done

**Comment by xilinxgitops** [comment]
retest this please

**Review by houlz0507 — APPROVED** [review]

File: src/runtime_src/core/common/drv/kds_core.c [hunk 2]
```diff
@@ -75,7 +75,7 @@ ssize_t show_kds_custat_raw(struct kds_sched *kds, char *buf)
 ssize_t show_kds_scustat_raw(struct kds_sched *kds, char *buf)
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/kds_core.c:75]: In all reality, this function's definition should pass in the size of the buffer instead of "assuming" that it is a PAGE_SIZE.

File: src/runtime_src/core/common/drv/kds_core.c [hunk 3]
```diff
@@ -93,13 +93,34 @@ ssize_t show_kds_scustat_raw(struct kds_sched *kds, char *buf)
...
 	}
 	mutex_unlock(&scu_mgmt->lock);
 
 	return sz;
 }
 
+ssize_t kds_sk_memstat(struct kds_sched *kds, struct sk_mem_stats *stat)
+{
+	struct kds_scu_mgmt *scu_mgmt = &kds->scu_mgmt;
+	struct sk_mem_stats *mem_stat;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/kds_core.c:109]: Uninitialized variable.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 4]
```diff
@@ -1082,6 +1082,7 @@ cu_stat(struct sched_cmd *cmd)
 	struct zocl_ert_dev *ert = zdev->ert;
 	struct sched_exec_core *exec = zdev->exec;
 	struct soft_krnl *sk = zdev->soft_kernel;
+	struct sk_mem_stats *mem_stat;
```
Comment (rozumx) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c:1085]: Multiple uninitialized variables.

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c [hunk 5]
```diff
@@ -147,6 +177,7 @@ zocl_sk_report_ioctl(struct drm_device *dev, void *data,
 	struct soft_cu *scu;
 	uint32_t cu_idx = args->cu_idx;
 	uint32_t *vaddr;
```
Comment (rozumx) [inline src/runtime_src/core/edge/drm/zocl/zocl_sk.c:179]: Uninitialized variable.

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c [hunk 6]
```diff
@@ -147,6 +177,7 @@ zocl_sk_report_ioctl(struct drm_device *dev, void *data,
 	struct soft_cu *scu;
```
Comment (rozumx) [inline src/runtime_src/core/edge/drm/zocl/zocl_sk.c:177]: Uninitialized variable.

File: src/runtime_src/core/include/ert.h [hunk 7]
```diff
@@ -625,6 +625,12 @@ uint32_t ert_base_addr = 0;
 # define ERT_INTC_CU_96_127_ADDR           0x3000
 #endif
 
+/**
+ * This defines ERT packet version which need to sync between Host and ERT
+ */
+#define ERT_CUSTAT_VERSION_0 		  0x51a10000
+#define ERT_CUSTAT_VERSION_1		  0x51a10001
```
Comment (rozumx) [inline src/runtime_src/core/include/ert.h:632]: Thank you for added these.  It does make the code much easier to read.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 8]
```diff
@@ -325,18 +325,25 @@ static inline void read_ert_stat(struct kds_command *xcmd)
...
 	/* New KDS handle FPGA CU statistic on host not ERT */
-	if (ecmd->data[0] != 0x51a10000)
+	if ((ecmd->data[0] != ERT_CUSTAT_VERSION_0) && 
+			(ecmd->data[0] != ERT_CUSTAT_VERSION_1))
 		return;
 
 	/* Only need PS kernel info, which is after FPGA CUs */
 	mutex_lock(&kds->scu_mgmt.lock);
 	/* Skip header and FPGA CU stats. off_idx points to PS kernel stats */
 	off_idx = 4 + num_cu;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:335]: What is so magical about the value 4.

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 9]
```diff
@@ -125,7 +125,7 @@ struct kds_scu_stat
       boost::char_separator<char> sep(",");
       tokenizer tokens(line, sep);
 
-      if (std::distance(tokens.begin(), tokens.end()) != 4)
+      if (std::distance(tokens.begin(), tokens.end()) != 7)
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:128]: Is it possible to have this value defined in a common place.  For example, in the kds_core.h file.  This way, this DR check would never need to be updated.

File: src/runtime_src/core/tools/common/ReportDynamicRegion.cpp [hunk 10]
```diff
@@ -316,26 +320,28 @@ ReportDynamicRegion::writeReport( const xrt_core::device* /*_pDevice*/,
...
+      _output << scuFmt % index++ %
+	      cu.get<std::string>("name") % cu.get<std::string>("base_address") %
+	      cu.get<std::string>("usage") % xrt_core::utils::parse_cu_status(status_val) %
+	      cu.get<std::string>("succ_return") % cu.get<std::string>("err_return") %
+	      cu.get<std::string>("crsh_return");
     }
   }
+  catch( std::exception const& e) {
+    _output << "ERROR: " <<  e.what() << std::endl;
+  }
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportDynamicRegion.cpp:343]: Looks to be a format indention issue.

File: src/runtime_src/core/tools/common/ReportMemory.cpp [hunk 11]
```diff
@@ -225,6 +244,20 @@ populate_memtopology(const xrt_core::device * device, const std::string& desc)
   }
   pt.add_child(std::string("board.memory.data_streams"), ptStream_array);
   pt.add_child(std::string("board.memory.memories"), ptMem_array );
+   
+  // Populate softkernel mem stats 
+  boost::property_tree::ptree ptSkMem;
+  std::stringstream ss_mem(mm_buf[map->m_count + 1]);
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportMemory.cpp:250]: Magic number "1", why is this needed?

It might be better to define a variable the represents the size prior creating the stringstream.  Again, looking at readability.

File: src/runtime_src/core/tools/common/ReportMemory.cpp [hunk 12]
```diff
@@ -389,6 +422,33 @@ ReportMemory::writeReport( const xrt_core::device* /*_pDevice*/,
...
 
+  bool sk_mem_is_present = _pt.get_child("mem_topology.board.memory.sk_memories",empty_ptree).size() > 0 ? true:false;
+  if (sk_mem_is_present) {
+    _output << std::endl;
+    _output << "  Soft Kernel Memory Status" << std::endl;
+    _output << boost::format("  %-20s%-16s%-16s%-16s\n") % "  HostBO Count" % "MapBO Count" % "UnmapBO Count" % "FreeBO Count";
+
+    try {
+      std::string hbo_count, mapbo_count, unmapbo_count, freebo_count;
+      for (auto& v : _pt.get_child("mem_topology.board.memory.sk_memories",empty_ptree)) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportMemory.cpp:433]: This for look seems overly complicated and possibly error prone. 

Is there a reason why you didn't consider the following:

```
auto & sk_memories_pt = _pt.get_child("mem_topology.board.memory.sk_memories",empty_ptree);
const std:string &hbo_count = sk_memories_pt.get<std::string>("sk_hbo","");
const std:string &mapbo_count = sk_memories_pt.get<std::string>("sk_mapbo","");
const std:string &unmapbo_count = sk_memories_pt.get<std::string>("sk_unmapbo","");
const std:string &freebo_count = sk_memories_pt.get<std::string>(sk_freebo,"");

```

**Review by rozumx — CHANGES_REQUESTED** [review]
Changes requested.

File: src/runtime_src/core/common/drv/kds_core.c (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c]: Fixed. Thanks.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: Fixed. Thanks.

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_sk.c]: Fixed. Thanks.

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_sk.c]: Fixed. Thanks.

File: src/runtime_src/core/include/ert.h (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/include/ert.h:632]: Thanks.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c]: Fixed. Thanks for this suggestion. Its looks good now and more readable.

File: src/runtime_src/core/pcie/linux/device_linux.cpp (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/linux/device_linux.cpp]: Fixed. Thanks.

File: src/runtime_src/core/tools/common/ReportDynamicRegion.cpp (hunk 10 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportDynamicRegion.cpp]: Fixed. Thanks.

File: src/runtime_src/core/tools/common/ReportMemory.cpp (hunk 11 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportMemory.cpp]: Fixed. Thanks.

File: src/runtime_src/core/tools/common/ReportMemory.cpp (hunk 12 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportMemory.cpp]: This suggestions are really helpful. Fixed. Thanks.

File: src/runtime_src/core/pcie/linux/device_linux.cpp (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/linux/device_linux.cpp]: The communication between driver and userspace are through sysfs file. To generalize this we need to update this value from driver to sysfs file and userspace has to read that value from sysfs file. 
This need some more code changes  from both driver and userspace side. As this PR targeted for U30 device, I don't want to make changes at this point of time. But for master we can make that changes. 
Created a CR to track this. https://jira.xilinx.com/browse/CR-1106063

Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/linux/device_linux.cpp]: Currently added a MACRO in the same file to avoid hard code the value.

File: src/runtime_src/core/common/drv/kds_core.c (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c:75]: Actually kernel default sysfs buffer size is one page, or PAGE_SIZE bytes. Hence, we need to assume or hard code the size in caller or inside this function.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 13]
```diff
@@ -523,6 +526,11 @@ int zocl_map_bo_ioctl(struct drm_device *dev,
 	args->offset = drm_vma_node_offset_addr(&gem_obj->vma_node);
 	zocl_describe(to_zocl_bo(gem_obj));
 
+	/* Update softkernel memory stats */
+	bo = to_zocl_bo(gem_obj);
+	if (bo->flags & ZOCL_BO_FLAGS_HOST_BO)
+        	zocl_update_sk_mem_stat(zdev, 1, ZOCL_MEM_STAT_TYPE_MAPBO);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c:532]: There is always add "1" when calling this function. Do you really need a function that add more than 1?

Maybe use zocl_sk_mem_stat_incr(zdev, ZOCL_MEM_STAT_TYPE_MAPBO); for better readability?

Try avoid constant value and if you have to do so, make the code self readable.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 14]
```diff
@@ -887,6 +895,8 @@ int zocl_get_hbo_ioctl(struct drm_device *dev, void *data,
 		return ret;
 	}
 
+	zocl_update_sk_mem_stat(zdev, 1, ZOCL_MEM_STAT_TYPE_HBO);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c:898]: same

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 15]
```diff
@@ -907,6 +918,9 @@ void zocl_free_host_bo(struct drm_gem_object *gem_obj)
 	drm_gem_object_release(gem_obj);
 
 	kfree(&zocl_bo->cma_base);
+			
+	/* Update softkernel memory stats */
+	zocl_update_sk_mem_stat(zdev, 1, ZOCL_MEM_STAT_TYPE_FREEBO);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c:923]: same

File: src/runtime_src/core/edge/drm/zocl/zocl_drv.c [hunk 16]
```diff
@@ -705,10 +705,23 @@ static int zocl_iommu_init(struct drm_zocl_dev *zdev,
...
 
+void zocl_drm_gem_vm_close(struct vm_area_struct *vma)
+{
+	struct drm_gem_object *obj = vma->vm_private_data;
+	struct drm_zocl_dev *zdev = obj->dev->dev_private;
+	struct drm_zocl_bo *bo = to_zocl_bo(obj);
+
+	/* Update softkernel memory stats */
+	if (bo->flags & ZOCL_BO_FLAGS_HOST_BO)
+		zocl_update_sk_mem_stat(zdev, 1, ZOCL_MEM_STAT_TYPE_UNMAPBO);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_drv.c:716]: same

File: src/runtime_src/core/include/ert.h [hunk 17]
```diff
@@ -625,6 +625,12 @@ uint32_t ert_base_addr = 0;
 # define ERT_INTC_CU_96_127_ADDR           0x3000
 #endif
 
+/**
+ * This defines ERT packet version which need to sync between Host and ERT
+ */
+#define ERT_CUSTAT_VERSION_0 		  0x51a10000
```
Comment (mamin506) [inline src/runtime_src/core/include/ert.h:631]: good change 👍

**Review by mamin506 — CHANGES_REQUESTED** [review]
Overall this looks good to me.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c (hunk 13 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Thanks for this suggestion. Updated the code accordingly.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c (hunk 14 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Fixed. Thanks.

File: src/runtime_src/core/edge/drm/zocl/zocl_drv.c (hunk 16 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_drv.c]: Fixed. Thanks.

File: src/runtime_src/core/include/ert.h (hunk 17 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/include/ert.h:631]: Thanks.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c (hunk 15 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Fixed. Thanks.

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c [hunk 18]
```diff
@@ -505,8 +506,10 @@ int zocl_map_bo_ioctl(struct drm_device *dev,
 		struct drm_file *filp)
 {
 	int ret = 0;
+	struct drm_zocl_dev *zdev = dev->dev_private;
 	struct drm_zocl_map_bo *args = data;
 	struct drm_gem_object *gem_obj;
+	struct drm_zocl_bo *bo;
```
Comment (rozumx) [inline src/runtime_src/core/edge/drm/zocl/zocl_bo.c:512]: Uninitialized pointers.

**Review by rozumx — APPROVED** [review]
Approved with minor requested changes.

**Review by larry9523 — APPROVED** [review]

File: src/runtime_src/core/edge/drm/zocl/zocl_bo.c (hunk 18 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_bo.c]: Fixed. Thanks

**Review by mamin506 — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]
